File: templates/minimal-provider.ts
/**
 * Minimal Provider — Hera Architecture Reference
 *
 * LLM provider abstraction. Pi supports 20+ providers through a unified API.
 * Based on packages/ai/src/providers/ and packages/ai/src/types.ts
 *
 * The key insight: all providers implement the same streaming interface.
 * This allows swapping providers without changing agent code.
 */

// ============================================================================
// Types
// ============================================================================

interface TextContent {
  type: "text";
  text: string;
}

interface ToolCallContent {
  type: "toolCall";
  id: string;
  name: string;
  arguments: Record<string, unknown>;
}

interface AssistantMessage {
  role: "assistant";
  content: (TextContent | ToolCallContent)[];
  model: string;
  provider: string;
  usage: {
    input: number;
    output: number;
    totalTokens: number;
  };
  stopReason: "end_turn" | "tool_use" | "error" | "aborted";
  errorMessage?: string;
  timestamp: number;
}

interface UserMessage {
  role: "user";
  content: TextContent[];
  timestamp: number;
}

interface ToolResultMessage {
  role: "toolResult";
  content: { toolCallId: string; content: TextContent[]; isError: boolean }[];
  timestamp: number;
}

type Message = UserMessage | AssistantMessage | ToolResultMessage;

interface Context {
  systemPrompt: string;
  messages: Message[];
  tools?: { name: string; description: string; parameters: Record<string, unknown> }[];
}

interface Model {
  id: string;
  name: string;
  provider: string;
  api: string;
  baseUrl: string;
  contextWindow: number;
  maxTokens: number;
}

interface StreamOptions {
  apiKey?: string;
  signal?: AbortSignal;
  temperature?: number;
  maxTokens?: number;
  headers?: Record<string, string>;
}

// ============================================================================
// Streaming Events
// ============================================================================

type StreamEvent =
  | { type: "start"; partial: AssistantMessage }
  | { type: "text_delta"; partial: AssistantMessage }
  | { type: "toolcall_delta"; partial: AssistantMessage }
  | { type: "done"; message: AssistantMessage }
  | { type: "error"; error: AssistantMessage };

// ============================================================================
// Event Stream (async iterator)
// ============================================================================

class EventStream<T, R> implements AsyncIterable<T> {
  private queue: T[] = [];
  private waiting: ((value: IteratorResult<T>) => void)[] = [];
  private done = false;
  private resultPromise: Promise<R>;
  private resolveResult!: (result: R) => void;

  constructor() {
    this.resultPromise = new Promise((resolve) => {
      this.resolveResult = resolve;
    });
  }

  push(event: T): void {
    if (this.done) return;
    const waiter = this.waiting.shift();
    if (waiter) {
      waiter({ value: event, done: false });
    } else {
      this.queue.push(event);
    }
  }

  end(result: R): void {
    this.done = true;
    this.resolveResult(result);
    while (this.waiting.length > 0) {
      this.waiting.shift()!({ value: undefined as any, done: true });
    }
  }

  async *[Symbol.asyncIterator](): AsyncIterator<T> {
    while (true) {
      if (this.queue.length > 0) {
        yield this.queue.shift()!;
      } else if (this.done) {
        return;
      } else {
        await new Promise<IteratorResult<T>>((resolve) => this.waiting.push(resolve));
      }
    }
  }

  result(): Promise<R> {
    return this.resultPromise;
  }
}

// ============================================================================
// Provider Interface
// ============================================================================

type ProviderHandler = (
  model: Model,
  context: Context,
  options: StreamOptions,
) => EventStream<StreamEvent, AssistantMessage>;

// ============================================================================
// OpenAI Provider Implementation
// ============================================================================

function createOpenAIProvider(): ProviderHandler {
  return (model, context, options) => {
    const stream = new EventStream<StreamEvent, AssistantMessage>();

    // Build request payload
    const payload = {
      model: model.id,
      messages: [
        { role: "system", content: context.systemPrompt },
        ...context.messages.map((m) => {
          if (m.role === "user") return { role: "user", content: m.content[0]?.text ?? "" };
          if (m.role === "assistant") return { role: "assistant", content: m.content[0]?.text ?? "" };
          return { role: "user", content: JSON.stringify(m.content) };
        }),
      ],
      tools: context.tools?.map((t) => ({
        type: "function",
        function: { name: t.name, description: t.description, parameters: t.parameters },
      })),
      stream: true,
      temperature: options.temperature ?? 0.7,
      max_tokens: options.maxTokens ?? 4096,
    };

    // In production, this would be an actual HTTP request
    // For this example, simulate the streaming response
    (async () => {
      const message: AssistantMessage = {
        role: "assistant",
        content: [{ type: "text", text: "This is a simulated response from OpenAI." }],
        model: model.id,
        provider: "openai",
        usage: { input: 100, output: 50, totalTokens: 150 },
        stopReason: "end_turn",
        timestamp: Date.now(),
      };

      stream.push({ type: "start", partial: message });
      await new Promise((r) => setTimeout(r, 50));
      stream.push({ type: "text_delta", partial: message });
      await new Promise((r) => setTimeout(r, 50));
      stream.push({ type: "done", message });
      stream.end(message);
    })();

    return stream;
  };
}

// ============================================================================
// Provider Registry
// ============================================================================

class ProviderRegistry {
  private providers = new Map<string, ProviderHandler>();

  register(api: string, handler: ProviderHandler): void {
    this.providers.set(api, handler);
  }

  get(api: string): ProviderHandler | undefined {
    return this.providers.get(api);
  }

  async stream(
    model: Model,
    context: Context,
    options: StreamOptions,
  ): Promise<EventStream<StreamEvent, AssistantMessage>> {
    const handler = this.providers.get(model.api);
    if (!handler) {
      throw new Error(`No provider registered for API: ${model.api}`);
    }
    return handler(model, context, options);
  }
}

// ============================================================================
// Example Usage
// ============================================================================

async function main() {
  const registry = new ProviderRegistry();

  // Register OpenAI provider
  registry.register("openai-chat", createOpenAIProvider());

  // Define model
  const model: Model = {
    id: "gpt-4",
    name: "GPT-4",
    provider: "openai",
    api: "openai-chat",
    baseUrl: "https://api.openai.com/v1",
    contextWindow: 128000,
    maxTokens: 4096,
  };

  // Create context
  const context: Context = {
    systemPrompt: "You are a helpful assistant.",
    messages: [
      { role: "user", content: [{ type: "text", text: "Hello!" }], timestamp: Date.now() },
    ],
  };

  // Stream response
  const stream = await registry.stream(model, context, { apiKey: "sk-..." });

  for await (const event of stream) {
    switch (event.type) {
      case "start":
        console.log("[Start] Streaming began");
        break;
      case "text_delta":
        console.log("[Text]", event.partial.content[0]?.text);
        break;
      case "done":
        console.log("[Done] Total tokens:", event.message.usage.totalTokens);
        break;
      case "error":
        console.error("[Error]", event.error.errorMessage);
        break;
    }
  }
}

main().catch(console.error);

File: templates/python/minimal_provider.py
"""
Minimal Provider — Hera Architecture Reference (Python)

LLM provider abstraction supporting OpenAI, Anthropic, and local models.
Handles retries, streaming, and multi-provider routing.

Based on Pi Agent's provider system (packages/agent/src/ai/).
"""

from __future__ import annotations

import asyncio
import json
import time
from abc import ABC, abstractmethod
from dataclasses import dataclass, field
from typing import Any, AsyncIterator


# ============================================================================
# Types
# ============================================================================

@dataclass
class Usage:
    input_tokens: int = 0
    output_tokens: int = 0
    total_tokens: int = 0
    cost_usd: float = 0.0
    duration_ms: float = 0.0


@dataclass
class StreamChunk:
    """Single chunk from a streaming LLM response."""
    text: str = ""
    tool_call_id: str | None = None
    tool_name: str | None = None
    tool_arguments: str = ""  # Partial JSON
    finish_reason: str | None = None


@dataclass
class ProviderConfig:
    """Configuration for an LLM provider."""
    name: str = ""
    api_key: str = ""
    base_url: str = ""
    model: str = ""
    max_tokens: int = 4096
    temperature: float = 0.7
    timeout: float = 60.0
    max_retries: int = 3
    cost_per_input_token: float = 0.0
    cost_per_output_token: float = 0.0


# ============================================================================
# Provider Interface
# ============================================================================

class LLMProvider(ABC):
    """Base class for LLM providers."""

    def __init__(self, config: ProviderConfig):
        self.config = config

    @abstractmethod
    async def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        """Non-streaming chat completion."""
        ...

    @abstractmethod
    async def chat_stream(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> AsyncIterator[StreamChunk]:
        """Streaming chat completion."""
        ...


# ============================================================================
# OpenAI Provider
# ============================================================================

class OpenAIProvider(LLMProvider):
    """OpenAI-compatible provider (works with OpenAI, vLLM, LiteLLM, etc.)."""

    async def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ImportError("pip install openai")

        client = AsyncOpenAI(
            api_key=self.config.api_key,
            base_url=self.config.base_url or None,
            timeout=self.config.timeout,
        )

        kwargs: dict[str, Any] = {
            "model": self.config.model,
            "messages": messages,
            "max_tokens": self.config.max_tokens,
            "temperature": self.config.temperature,
        }

        if tools:
            kwargs["tools"] = tools
            kwargs["tool_choice"] = "auto"

        start = time.time()
        response = await client.chat.completions.create(**kwargs)
        duration = (time.time() - start) * 1000

        choice = response.choices[0]
        usage = Usage(
            input_tokens=response.usage.prompt_tokens if response.usage else 0,
            output_tokens=response.usage.completion_tokens if response.usage else 0,
            total_tokens=response.usage.total_tokens if response.usage else 0,
            duration_ms=duration,
        )

        # Calculate cost
        usage.cost_usd = (
            usage.input_tokens * self.config.cost_per_input_token
            + usage.output_tokens * self.config.cost_per_output_token
        )

        return {
            "content": choice.message.content or "",
            "tool_calls": [
                {
                    "id": tc.id,
                    "name": tc.function.name,
                    "arguments": json.loads(tc.function.arguments),
                }
                for tc in (choice.message.tool_calls or [])
            ],
            "finish_reason": choice.finish_reason,
            "usage": usage,
        }

    async def chat_stream(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> AsyncIterator[StreamChunk]:
        try:
            from openai import AsyncOpenAI
        except ImportError:
            raise ImportError("pip install openai")

        client = AsyncOpenAI(
            api_key=self.config.api_key,
            base_url=self.config.base_url or None,
        )

        kwargs: dict[str, Any] = {
            "model": self.config.model,
            "messages": messages,
            "max_tokens": self.config.max_tokens,
            "temperature": self.config.temperature,
            "stream": True,
        }

        if tools:
            kwargs["tools"] = tools
            kwargs["tool_choice"] = "auto"

        stream = await client.chat.completions.create(**kwargs)

        current_tool_id: str | None = None
        current_tool_name: str | None = None
        current_tool_args = ""

        async for chunk in stream:
            delta = chunk.choices[0].delta if chunk.choices else None
            finish = chunk.choices[0].finish_reason if chunk.choices else None

            if delta:
                # Text content
                if delta.content:
                    yield StreamChunk(text=delta.content)

                # Tool calls
                if delta.tool_calls:
                    for tc in delta.tool_calls:
                        if tc.id:
                            current_tool_id = tc.id
                            current_tool_name = tc.function.name if tc.function else None
                            current_tool_args = ""

                        if tc.function and tc.function.arguments:
                            current_tool_args += tc.function.arguments

                        # Try to parse complete tool call
                        try:
                            json.loads(current_tool_args)
                            yield StreamChunk(
                                tool_call_id=current_tool_id,
                                tool_name=current_tool_name,
                                tool_arguments=current_tool_args,
                            )
                            current_tool_id = None
                            current_tool_name = None
                            current_tool_args = ""
                        except json.JSONDecodeError:
                            pass  # Incomplete JSON, continue accumulating

            if finish:
                yield StreamChunk(finish_reason=finish)


# ============================================================================
# Anthropic Provider
# ============================================================================

class AnthropicProvider(LLMProvider):
    """Anthropic Claude provider."""

    async def chat(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> dict[str, Any]:
        try:
            from anthropic import AsyncAnthropic
        except ImportError:
            raise ImportError("pip install anthropic")

        client = AsyncAnthropic(api_key=self.config.api_key, timeout=self.config.timeout)

        # Anthropic uses system as separate param
        system_msg = ""
        chat_messages = []
        for msg in messages:
            if msg["role"] == "system":
                system_msg = msg["content"]
            else:
                chat_messages.append(msg)

        kwargs: dict[str, Any] = {
            "model": self.config.model,
            "messages": chat_messages,
            "max_tokens": self.config.max_tokens,
        }

        if system_msg:
            kwargs["system"] = system_msg

        if tools:
            kwargs["tools"] = tools

        start = time.time()
        response = await client.messages.create(**kwargs)
        duration = (time.time() - start) * 1000

        usage = Usage(
            input_tokens=response.usage.input_tokens,
            output_tokens=response.usage.output_tokens,
            duration_ms=duration,
        )

        return {
            "content": response.content[0].text if response.content else "",
            "tool_calls": [],
            "finish_reason": response.stop_reason,
            "usage": usage,
        }

    async def chat_stream(
        self,
        messages: list[dict[str, Any]],
        tools: list[dict[str, Any]] | None = None,
    ) -> AsyncIterator[StreamChunk]:
        try:
            from anthropic import AsyncAnthropic
        except ImportError:
            raise ImportError("pip install anthropic")

        client = AsyncAnthropic(api_key=self.config.api_key)

        system_msg = ""
        chat_messages = []
        for msg in messages:
            if msg["role"] == "system":
                system_msg = msg["content"]
            else:
                chat_messages.append(msg)

        kwargs: dict[str, Any] = {
            "model": self.config.model,
            "messages": chat_messages,
            "max_tokens": self.config.max_tokens,
        }

        if system_msg:
            kwargs["system"] = system_msg

        if tools:
            kwargs["tools"] = tools

        async with client.messages.stream(**kwargs) as stream:
            async for text in stream.text_stream:
                yield StreamChunk(text=text)

            yield StreamChunk(finish_reason="end_turn")


# ============================================================================
# Example Usage
# ============================================================================

async def main():
    config = ProviderConfig(
        name="openai",
        api_key="sk-...",
        model="gpt-4o",
        max_tokens=4096,
    )

    provider = OpenAIProvider(config)

    messages = [
        {"role": "system", "content": "You are a helpful coding assistant."},
        {"role": "user", "content": "Write a hello world in Python"},
    ]

    # Non-streaming
    result = await provider.chat(messages)
    print(f"Response: {result['content']}")
    print(f"Tokens: {result['usage'].total_tokens}")
    print(f"Cost: ${result['usage'].cost_usd:.4f}")

    # Streaming
    print("\n--- Streaming ---")
    async for chunk in provider.chat_stream(messages):
        if chunk.text:
            print(chunk.text, end="", flush=True)
        if chunk.finish_reason:
            print(f"\n[Finish: {chunk.finish_reason}]")


if __name__ == "__main__":
    asyncio.run(main())
